core-primitives/stf-state-handler/src/in_memory_state_file_io.rs

```diff
@@ -0,0 +1,378 @@
+		let state_entry = states_for_shard
+			.entry(state_id)
+			.or_insert_with(|| self.generate_state_entry(State::default()));
+		Ok(state_entry.0)
+	}
+
+	fn write(
+		&self,
+		shard_identifier: &ShardIdentifier,
+		state_id: StateId,
+		state: Self::StateType,
+	) -> Result<Self::HashType> {
+		let mut directory_lock =
+			self.emulated_shard_directory.write().map_err(|_| Error::LockPoisoning)?;
+
+		let states_for_shard = directory_lock
+			.entry(*shard_identifier)
+			.or_insert_with(|| self.default_states_map(state_id));
+
+		let state_hash = self.compute_state_hash(&state);
+		*states_for_shard
+			.entry(state_id)
+			.or_insert_with(|| self.generate_default_state_entry()) = (state_hash, state);
+
+		Ok(state_hash)
+	}
+
+	fn remove(&self, shard_identifier: &ShardIdentifier, state_id: StateId) -> Result<()> {
+		let mut directory_lock =
+			self.emulated_shard_directory.write().map_err(|_| Error::LockPoisoning)?;
+
+		let states_for_shard = directory_lock
+			.get_mut(shard_identifier)
+			.ok_or_else(|| Error::InvalidShard(*shard_identifier))?;
+
+		states_for_shard
+			.remove(&state_id)
+			.ok_or(Error::InvalidStateId(state_id))
+			.map(|_| {})
+	}
+
+	fn shard_exists(&self, shard_identifier: &ShardIdentifier) -> bool {
+		let directory_lock = self.emulated_shard_directory.read().unwrap();
+		directory_lock.contains_key(shard_identifier)
+	}
+
+	fn list_shards(&self) -> Result<Vec<ShardIdentifier>> {
+		let directory_lock =
+			self.emulated_shard_directory.read().map_err(|_| Error::LockPoisoning)?;
+		Ok(directory_lock.keys().copied().collect())
+	}
+
+	fn list_state_ids_for_shard(&self, shard_identifier: &ShardIdentifier) -> Result<Vec<StateId>> {
+		let directory_lock =
+			self.emulated_shard_directory.read().map_err(|_| Error::LockPoisoning)?;
+		let shard_directory = directory_lock
+			.get(shard_identifier)
+			.ok_or_else(|| Error::InvalidShard(*shard_identifier))?;
+		Ok(shard_directory.keys().cloned().collect())
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use std::{assert_matches::assert_matches, collections::hash_map::DefaultHasher};
+
+	type TestState = u64;
+	type TestStateFileIo = InMemoryStateFileIo<TestState, DefaultHasher>;
+
+	#[test]
+	fn shard_directory_is_empty_after_initialization() {
+		let state_file_io = create_empty_in_memory_state_file_io();
+		assert!(state_file_io.list_shards().unwrap().is_empty());
+	}
+
+	#[test]
+	fn load_on_empty_directory_and_shard_returns_error() {
+		let state_file_io = create_empty_in_memory_state_file_io();
+
+		assert_matches!(
+			state_file_io.load(&ShardIdentifier::random(), 1234),
+			Err(Error::InvalidShard(_))
+		);
+	}
+
+	#[test]
+	fn initialize_with_shard_creates_empty_directory() {
+		let shard = ShardIdentifier::random();
+		let state_file_io = create_in_memory_state_file_io(&[shard]);
+
+		assert!(state_file_io.list_state_ids_for_shard(&shard).unwrap().is_empty());
+		assert!(state_file_io.list_state_ids_for_shard(&ShardIdentifier::random()).is_err());
```

Comment:
That's true. The reason why in this case it's fine, IMO, is because we don't care at all about the shard identifier, it can have any value it wants, it's not relevant for any of the test conditions.

Reply:
As long as they are different, agreed. But the chance is so low, I guess it's fine

Reply:
Ah I didn't look at the whole test - no I think you have a point. Because we're comparing two shard identifiers that are both random, I agree that it's better to have pre-defined values here 👍